terraform-module-standards.md

```diff
@@ -0,0 +1,340 @@
+
+Workspaces can be listed with the `workspace` subcommand:
+```
+> terraform workspace list
+* default
+```
+
+To create a new workspace:
+```
+> terraform workspace new prod
+Created and switched to workspace "prod"!
+
+You're now on a new, empty workspace. Workspaces isolate their state,
+so if you run "terraform plan" Terraform will not see any existing state
+for this configuration.
+```
+Workspaces isolate their state by creating an additional state file in the state backend for each workspace.
+
+You must select a workspace before issuing terraform commands to that workspace.
+```
+> terraform workspace select prod
+Switched to workspace "prod".
+```
+
+Once you have switched to a workspace, all terraform commands work in the same manner you are accustomed to. A `tfvars` should be kept for each workspace to encapsulate the inputs for a given environment. `tfvars` can be stored in remote storage like GCS or S3 or encrypted and stored alongside code in a source code management system using KMS or a tool like git-crypt or git-secret.
+```
+> terraform apply -var-file=prod.tfvars
+```
+
+The `tfvars` file will drive the configuration of a specific environment. To enable/disable resources per environment, it is common to use the `count` attribute.
+
+Example: prod.tfvars
+```
+replica_count = 5
+```
+
+Example: staging.tfvars
+```
+replica_count = 1
+```
+
+Example: dev.tfvars
+```
+replica_count = 0
+```
+
+Example: main.tf
+```
+resource "google_sql_database_instance" "primary" {
+  name = "${terraform.workspace}-primary" //... 
+}
+
+resource "google_sql_database_instance" "replica" { 
+  name = "${terraform.workspace}-replica-${count.index}"
+  count = "${var.replica_count} //...
+}
+```
+
+### Outputs
+* Information from a root module which other root modules may depend on must be exported as outputs.
+* Root module outputs can be referenced using remote state.
+
+Publishing outputs with remote states:
+* Make sure to re-output nested module outputs that are useful as remote state.
+  * Only root module-level outputs can be referenced from other Terraform environments/applications.
+* Information related to a service’s endpoints should be expo	ed to remote state to allow use by other dependent apps for configuration.
+
+## Versioning
+### Terraform
+Terraform v0.12 is a significant release that includes some backwards incompatibilities and significant changes with each new version. Once a Terraform config has been upgraded to Terraform 0.12, it should be pinned to the latest release it has been tested with.
+```
+terraform {
+  required_version = "~> 0.12.16"
+}
+```
+
+### Equinix provider
+In Terraform root configurations, you should pin the Equinix provider to a known good **minor** version. This will allow automatic upgrade to new patch releases while still keeping a solid target. You should make updating the version pin a regular practice:
+```
+provider "equinix" {
+  source = "equinix/equinix"
+  version = "~> 1.11.0-alpha.2"
```

Comment:
If we are calling out `alpha` builds, we should call out the special behavior around these releases: https://developer.hashicorp.com/terraform/language/expressions/version-constraints#version-constraint-behavior

Reply:
Thanks! This has been addressed on commit 029c82e089218dbf238fc2d3eaf5fe0b2930ce15